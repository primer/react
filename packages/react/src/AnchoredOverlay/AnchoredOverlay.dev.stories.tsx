import type {Meta} from '@storybook/react'
import React, {useState} from 'react'

import {Button} from '../Button'
import {AnchoredOverlay} from '.'
import {Stack} from '../Stack'
import {Dialog, Spinner} from '..'

const meta = {
  title: 'Components/AnchoredOverlay/Dev',
  component: AnchoredOverlay,
} satisfies Meta<typeof AnchoredOverlay>

export default meta

export const RepositionAfterContentGrows = () => {
  const [open, setOpen] = useState(false)

  const [loading, setLoading] = useState(true)

  React.useEffect(() => {
    window.setTimeout(() => {
      if (open) setLoading(false)
    }, 2000)
  }, [open])

  return (
    <Stack direction="vertical" justify="space-between" style={{height: 'calc(100vh - 200px)'}}>
      <div>
        What to expect:
        <ul>
          <li>The anchored overlay should open below the anchor (default position)</li>
          <li>After 2000ms, the amount of content in the overlay grows</li>
          <li>the overlay should reposition itself above the anchor so that it stays inside the window</li>
        </ul>
      </div>
      <AnchoredOverlay
        renderAnchor={props => (
          <Button {...props} sx={{width: 'fit-content'}}>
            Button
          </Button>
        )}
        open={open}
        onOpen={() => setOpen(true)}
        onClose={() => {
          setOpen(false)
          setLoading(true)
        }}
      >
        {loading ? (
          <>
            <Spinner />
            loading for 2000ms
          </>
        ) : (
          <div style={{height: '300px'}}>content with 300px height</div>
        )}
      </AnchoredOverlay>
    </Stack>
  )
}

export const RepositionAfterContentGrowsWithinDialog = () => {
  const [open, setOpen] = useState(false)

  const [loading, setLoading] = useState(true)

  React.useEffect(() => {
    window.setTimeout(() => {
      if (open) setLoading(false)
    }, 2000)
  }, [open])

  return (
    <Dialog onClose={() => {}}>
      <Stack direction="vertical" justify="space-between" style={{height: 'calc(100vh - 300px)'}}>
        <div>
          What to expect:
          <ul>
            <li>The anchored overlay should open below the anchor (default position)</li>
            <li>After 2000ms, the amount of content in the overlay grows</li>
            <li>the overlay should reposition itself above the anchor so that it stays inside the window</li>
          </ul>
        </div>
        <AnchoredOverlay
          renderAnchor={props => (
            <Button {...props} sx={{width: 'fit-content'}}>
              Button
            </Button>
          )}
          open={open}
          onOpen={() => setOpen(true)}
          onClose={() => {
            setOpen(false)
            setLoading(true)
          }}
        >
          {loading ? (
            <>
              <Spinner />
              loading for 2000ms
            </>
          ) : (
            <div style={{height: '300px'}}>content with 300px height</div>
          )}
        </AnchoredOverlay>
      </Stack>
    </Dialog>
  )
}
