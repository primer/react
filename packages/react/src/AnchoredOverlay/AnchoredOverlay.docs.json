{
  "id": "anchored_overlay",
  "name": "AnchoredOverlay",
  "status": "alpha",
  "a11yReviewed": false,
  "stories": [],
  "importPath": "@primer/react",
  "props": [
    {
      "name": "open",
      "type": "boolean",
      "defaultValue": "false",
      "description": "Determines whether the overlay portion of the component should be shown or not."
    },
    {
      "name": "onOpen",
      "type": "(gesture: 'anchor-click' | 'anchor-key-press') => unknown",
      "defaultValue": "",
      "description": "A callback that is called whenever the overlay is currently closed and an \"open gesture\" is detected."
    },
    {
      "name": "onClose",
      "type": "(gesture: 'anchor-click' | 'click-outside' | 'escape') => unknown",
      "defaultValue": "",
      "description": "A callback that is called whenever the overlay is currently open and a \"close gesture\" is detected."
    },
    {
      "name": "renderAnchor",
      "type": "<T extends React.HTMLAttributes<HTMLElement>>(props: T) => JSX.Element",
      "defaultValue": "",
      "description": "A custom function component used to render the anchor element. When renderAnchor is null, an anchorRef is required."
    },
    {
      "name": "anchorRef",
      "type": "React.RefObject<HTMLElement>",
      "defaultValue": "",
      "description": "An override to the internal `renderAnchor` ref that will be used to position the overlay. When `renderAnchor` is `null`, this can be used to make an anchor that is detached from ` AnchoredOverlay `."
    },
    {
      "name": "anchorId",
      "type": "string",
      "defaultValue": "",
      "description": "An override to the internal id that will be passed to the anchor."
    },
    {
      "name": "side",
      "type": "| 'inside-top' | 'inside-bottom' | 'inside-left' | 'inside-right' | 'inside-center' | 'outside-top' | 'outside-bottom' | 'outside-left' | 'outside-right'",
      "defaultValue": "'outside-bottom'",
      "description": ""
    },
    {
      "name": "align",
      "type": "'start' | 'center' | 'end'",
      "defaultValue": "'start'",
      "description": ""
    },
    {
      "name": "alignmentOffset",
      "type": "number",
      "defaultValue": "4",
      "description": "An additional offset, in pixels, to move the floating element from the aligning edge. Positive values move the floating element in the direction of center-alignment. Negative values move the floating element away from center-alignment. When align is `'center`, positive offsets move the floating element right (top or bottom anchor side) or down (left or right anchor side). If using outside positioning, or if `align` is set to `'center'`, this defaults to `0` instead of `4`."
    },
    {
      "name": "anchorOffset",
      "type": "number",
      "defaultValue": "4",
      "description": "The number of pixels between the anchor edge and the floating element. Positive values move the floating element farther from the anchor element (for outside positioning) or further inside the anchor element (for inside positioning). Negative values have the opposite effect. If `side` is set to `'inside-center'`, this defaults to `0` instead of `4`."
    },
    {
      "name": "overlayProps",
      "type": "Partial<OverlayProps>",
      "defaultValue": "",
      "description": "Props to be spread on the internal `Overlay` component."
    },
    {
      "name": "focusTrapSettings",
      "type": "Partial<FocusTrapHookSettings>",
      "defaultValue": "",
      "description": "Settings to apply to the focus trap on the internal `Overlay` component."
    },
    {
      "name": "focusZoneSettings",
      "type": "Partial<FocusZoneHookSettings>",
      "defaultValue": "",
      "description": "Settings to apply to the focus zone on the internal `Overlay` component."
    },
    {
      "name": "className",
      "type": "string",
      "defaultValue": "",
      "description": "Class name for custom styling."
    }
  ]
}
