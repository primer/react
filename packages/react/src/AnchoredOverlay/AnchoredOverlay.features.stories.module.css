.HeaderAndLayout {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  padding: var(--base-size-24);
  width: 100%;
}

.ScrollingRegion {
  position: absolute;
  inset: 96px 24px 24px;
  overflow: scroll;
  /* stylelint-disable-next-line color-named */
  background-color: powderblue;
}

.PortalRootRegion {
  position: absolute;
  top: 0;
  left: 0;
}

.FlexColFill {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
}

.PlaygroundCell {
  margin: var(--base-size-8);
}

.Overlay {
  min-width: 320px;
}

.UserName {
  font-weight: var(--base-text-weight-medium, 500);
  color: var(--fgColor-default);
}

.UserMeta {
  color: var(--fgColor-muted);
  margin-left: var(--base-size-4);
}

.Bio {
  font-size: var(--text-body-size-medium);
  line-height: var(--text-body-lineHeight-medium);
}

.MetaMuted {
  font-size: var(--text-body-size-small);
  line-height: var(--text-body-lineHeight-small);
  color: var(--fgColor-muted);
  margin-left: var(--base-size-4);
}
