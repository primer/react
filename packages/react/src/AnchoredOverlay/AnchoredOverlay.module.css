.ResponsiveCloseButtonContainer {
  position: relative;
}

.ResponsiveCloseButton {
  position: absolute;
  top: var(--base-size-8);
  right: var(--base-size-8);
  display: none;

  @media screen and (--viewportRange-narrow) {
    display: inline-grid;
  }
}
