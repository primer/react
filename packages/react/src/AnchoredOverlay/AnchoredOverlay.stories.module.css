.TextMutedWithMargin {
  color: var(--fgColor-muted);
  margin-left: var(--base-size-4);
}

.TextSmallMutedWithMargin {
  font-size: var(--text-body-size-small);
  color: var(--fgColor-muted);
  margin-left: var(--base-size-4);
}
