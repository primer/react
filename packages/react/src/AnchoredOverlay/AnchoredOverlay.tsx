import React, {useCallback, useEffect} from 'react'
import type {OverlayProps} from '../Overlay'
import Overlay from '../Overlay'
import type {FocusTrapHookSettings} from '../hooks/useFocusTrap'
import {useFocusTrap} from '../hooks/useFocusTrap'
import type {FocusZoneHookSettings} from '../hooks/useFocusZone'
import {useFocusZone} from '../hooks/useFocusZone'
import {useAnchoredPosition, useProvidedRefOrCreate, useRenderForcingRef} from '../hooks'
import {useId} from '../hooks/useId'
import type {PositionSettings} from '@primer/behaviors'

interface AnchoredOverlayPropsWithAnchor {
  /**
   * A custom function component used to render the anchor element.
   * Will receive the selected text as `children` prop when an item is activated.
   */
  renderAnchor: <T extends React.HTMLAttributes<HTMLElement>>(props: T) => JSX.Element

  /**
   * An override to the internal ref that will be spread on to the renderAnchor
   */
  anchorRef?: React.RefObject<HTMLElement>

  /**
   * An override to the internal id that will be spread on to the renderAnchor
   */
  anchorId?: string
}

interface AnchoredOverlayPropsWithoutAnchor {
  /**
   * A custom function component used to render the anchor element.
   * When renderAnchor is null, an anchorRef is required.
   */
  renderAnchor: null

  /**
   * An override to the internal renderAnchor ref that will be used to position the overlay.
   * When renderAnchor is null this can be used to make an anchor that is detached from ActionMenu.
   */
  anchorRef: React.RefObject<HTMLElement>
  /**
   * An override to the internal id that will be spread on to the renderAnchor
   */
  anchorId?: string
}

export type AnchoredOverlayWrapperAnchorProps =
  | Partial<AnchoredOverlayPropsWithAnchor>
  | AnchoredOverlayPropsWithoutAnchor

interface AnchoredOverlayBaseProps extends Pick<OverlayProps, 'height' | 'width'> {
  /**
   * Determines whether the overlay portion of the component should be shown or not
   */
  open: boolean

  /**
   * A callback which is called whenever the overlay is currently closed and an "open gesture" is detected.
   */
  onOpen?: (gesture: 'anchor-click' | 'anchor-key-press', event?: React.KeyboardEvent<HTMLElement>) => unknown

  /**
   * A callback which is called whenever the overlay is currently open and a "close gesture" is detected.
   */
  onClose?: (gesture: 'anchor-click' | 'click-outside' | 'escape') => unknown

  /**
   * Props to be spread on the internal `Overlay` component.
   */
  overlayProps?: Partial<OverlayProps>

  /**
   * Settings to apply to the Focus Zone on the internal `Overlay` component.
   */
  focusTrapSettings?: Partial<FocusTrapHookSettings>

  /**
   * Settings to apply to the Focus Zone on the internal `Overlay` component.
   */
  focusZoneSettings?: Partial<FocusZoneHookSettings>
}

export type AnchoredOverlayProps = AnchoredOverlayBaseProps &
  (AnchoredOverlayPropsWithAnchor | AnchoredOverlayPropsWithoutAnchor) &
  Partial<Pick<PositionSettings, 'align' | 'side' | 'anchorOffset' | 'alignmentOffset'>>

/**
 * An `AnchoredOverlay` provides an anchor that will open a floating overlay positioned relative to the anchor.
 * The overlay can be opened and navigated using keyboard or mouse.
 */
export const AnchoredOverlay: React.FC<React.PropsWithChildren<AnchoredOverlayProps>> = ({
  renderAnchor,
  anchorRef: externalAnchorRef,
  anchorId: externalAnchorId,
  children,
  open,
  onOpen,
  onClose,
  height,
  width,
  overlayProps,
  focusTrapSettings,
  focusZoneSettings,
  side = 'outside-bottom',
  align = 'start',
  alignmentOffset,
  anchorOffset,
}) => {
  const anchorRef = useProvidedRefOrCreate(externalAnchorRef)
  const [overlayRef, updateOverlayRef] = useRenderForcingRef<HTMLDivElement>()
  const anchorId = useId(externalAnchorId)

  const onClickOutside = useCallback(() => onClose?.('click-outside'), [onClose])
  const onEscape = useCallback(() => onClose?.('escape'), [onClose])

  const onAnchorKeyDown = useCallback(
    (event: React.KeyboardEvent<HTMLElement>) => {
      if (!event.defaultPrevented) {
        if (!open && ['ArrowDown', 'ArrowUp', ' ', 'Enter'].includes(event.key)) {
          onOpen?.('anchor-key-press', event)
          event.preventDefault()
        }
      }
    },
    [open, onOpen],
  )
  const onAnchorClick = useCallback(
    (event: React.MouseEvent<HTMLElement>) => {
      if (event.defaultPrevented || event.button !== 0) {
        return
      }
      if (!open) {
        onOpen?.('anchor-click')
      } else {
        onClose?.('anchor-click')
      }
    },
    [open, onOpen, onClose],
  )

  const {position} = useAnchoredPosition(
    {
      anchorElementRef: anchorRef,
      floatingElementRef: overlayRef,
      side,
      align,
      alignmentOffset,
      anchorOffset,
    },
    [overlayRef.current],
  )

  useEffect(() => {
    // ensure overlay ref gets cleared when closed, so position can reset between closing/re-opening
    if (!open && overlayRef.current) {
      updateOverlayRef(null)
    }
  }, [open, overlayRef, updateOverlayRef])

  useFocusZone({
    containerRef: overlayRef,
    disabled: !open || !position,
    ...focusZoneSettings,
  })
  useFocusTrap({containerRef: overlayRef, disabled: !open || !position, ...focusTrapSettings})

  return (
    <>
      {renderAnchor &&
        renderAnchor({
          ref: anchorRef,
          id: anchorId,
          'aria-haspopup': 'true',
          'aria-expanded': open,
          tabIndex: 0,
          onClick: onAnchorClick,
          onKeyDown: onAnchorKeyDown,
        })}
      {open ? (
        <Overlay
          returnFocusRef={anchorRef}
          onClickOutside={onClickOutside}
          ignoreClickRefs={[anchorRef]}
          onEscape={onEscape}
          ref={updateOverlayRef}
          role="none"
          visibility={position ? 'visible' : 'hidden'}
          height={height}
          width={width}
          top={position?.top || 0}
          left={position?.left || 0}
          anchorSide={position?.anchorSide}
          {...overlayProps}
        >
          {children}
        </Overlay>
      ) : null}
    </>
  )
}

AnchoredOverlay.displayName = 'AnchoredOverlay'
