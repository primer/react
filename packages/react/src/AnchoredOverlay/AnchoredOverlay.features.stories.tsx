import React, {useEffect, useRef} from 'react'
import type {Args, Meta} from '@storybook/react'

import {Box} from '..'
import {AnchoredOverlay} from '../AnchoredOverlay'
import Heading from '../Heading'
import {registerPortalRoot} from '../Portal'
import {Playground} from './AnchoredOverlay.stories'

export default {
  title: 'Components/AnchoredOverlay/Features',
  component: AnchoredOverlay,
} as Meta

const HeaderAndLayout = ({children}: {children: JSX.Element}) => {
  const scrollingElementRef = useRef<HTMLDivElement>(null)
  useEffect(() => {
    if (scrollingElementRef.current) {
      registerPortalRoot(scrollingElementRef.current, 'scrollingPortal')
    }
  }, [scrollingElementRef])
  return (
    <Box position="absolute" top={0} right={0} bottom={0} left={0} padding={4} backgroundColor="lavenderblush">
      <Heading>Header or some such</Heading>
      <Box position="absolute" top={10} right={4} bottom={4} left={4} overflow="scroll" backgroundColor="powderblue">
        {children}
        <Box ref={scrollingElementRef} position="absolute" top={0} left={0} />
      </Box>
    </Box>
  )
}

export const PortalInsideScrollingElement = (args: Args) => {
  const rows = 20
  const columns = 10
  return (
    <HeaderAndLayout>
      <table>
        <tbody>
          {Array(rows)
            .fill(null)
            .map((_, i) => (
              <tr key={i}>
                {Array(columns)
                  .fill(null)
                  .map((_1, j) => (
                    <td key={`${i}${j}`}>
                      <Box m={2}>
                        <Playground {...{...args, portalContainerName: 'scrollingPortal'}} />
                      </Box>
                    </td>
                  ))}
              </tr>
            ))}
        </tbody>
      </table>
    </HeaderAndLayout>
  )
}
