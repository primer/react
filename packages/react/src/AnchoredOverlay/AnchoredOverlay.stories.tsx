import React, {useState} from 'react'
import type {Args, Meta} from '@storybook/react'

import {Box} from '..'
import {AnchoredOverlay} from '../AnchoredOverlay'
import {Button} from '../Button'
import {action} from '@storybook/addon-actions'

export default {
  title: 'Components/AnchoredOverlay',
  component: AnchoredOverlay,
} as Meta

export const Default = () => {
  const [open, setOpen] = useState(false)

  return (
    <AnchoredOverlay
      open={open}
      onOpen={() => setOpen(true)}
      onClose={() => setOpen(false)}
      renderAnchor={props => <Button {...props}>Button</Button>}
    >
      <Box width="100%" height="100%" backgroundColor="thistle" display="flex" flexDirection="column">
        <img src={`//placekitten.com/200/300`} alt="kitten" />
      </Box>
    </AnchoredOverlay>
  )
}

export const Playground = (args: Args) => {
  const [open, setOpen] = useState(false)

  return (
    <AnchoredOverlay
      open={open}
      onOpen={() => {
        setOpen(true)
        action('onOpen')()
      }}
      onClose={() => {
        setOpen(false)
        action('onClose')()
      }}
      width={args.width}
      height={args.height}
      renderAnchor={props => <Button {...props}>Button</Button>}
      overlayProps={args.portalContainerName}
      side={args.side}
    >
      <Box width="100%" height="100%" backgroundColor="thistle" display="flex" flexDirection="column">
        <img src={`//placekitten.com/200/300`} alt="kitten" />
      </Box>
    </AnchoredOverlay>
  )
}
Playground.args = {
  width: 'small',
  height: 'auto',
  portalContainerName: 'Portal name',
  side: 'outside-bottom',
}
Playground.argTypes = {
  width: {
    type: {
      name: 'enum',
      value: ['small', 'medium', 'large', 'xlarge', 'xxlarge', 'auto'],
    },
    portalContainerName: {type: 'string'},
  },
  height: {
    type: {
      name: 'enum',
      value: ['xsmall', 'small', 'medium', 'large', 'xlarge', 'auto', 'initial'],
    },
    portalContainerName: {type: 'string'},
  },
  side: {
    type: {
      name: 'enum',
      value: [
        'inside-top',
        'inside-bottom',
        'inside-left',
        'inside-right',
        'inside-center',
        'outside-top',
        'outside-bottom',
        'outside-left',
        'outside-right',
      ],
    },
  },
  open: {
    control: false,
  },
}
