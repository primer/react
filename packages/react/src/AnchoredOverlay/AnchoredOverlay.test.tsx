import React, {useCallback, useState} from 'react'
import {describe, expect, it, vi} from 'vitest'
import {render, fireEvent} from '@testing-library/react'
import {AnchoredOverlay} from '../AnchoredOverlay'
import {Button} from '../Button'
import theme from '../theme'
import BaseStyles from '../BaseStyles'
import {ThemeProvider} from '../ThemeProvider'
import type {AnchorPosition} from '@primer/behaviors'
type TestComponentSettings = {
  initiallyOpen?: boolean
  onOpenCallback?: (gesture: string) => void
  onCloseCallback?: (gesture: string) => void
  onPositionChange?: ({position}: {position: AnchorPosition}) => void
}

const AnchoredOverlayTestComponent = ({
  initiallyOpen = false,
  onOpenCallback,
  onCloseCallback,
  onPositionChange,
}: TestComponentSettings = {}) => {
  const [open, setOpen] = useState(initiallyOpen)
  const onOpen = useCallback(
    (gesture: string) => {
      setOpen(true)
      onOpenCallback?.(gesture)
    },
    [onOpenCallback],
  )
  const onClose = useCallback(
    (gesture: string) => {
      setOpen(true)
      onCloseCallback?.(gesture)
    },
    [onCloseCallback],
  )
  return (
    <ThemeProvider theme={theme}>
      <BaseStyles>
        <AnchoredOverlay
          open={open}
          onOpen={onOpen}
          onClose={onClose}
          renderAnchor={props => <Button {...props}>Anchor Button</Button>}
          onPositionChange={onPositionChange}
        >
          <button type="button">Focusable Child</button>
        </AnchoredOverlay>
      </BaseStyles>
    </ThemeProvider>
  )
}

describe('AnchoredOverlay', () => {
  it('should call onOpen when the anchor is clicked', () => {
    const mockOpenCallback = vi.fn()
    const mockCloseCallback = vi.fn()
    const anchoredOverlay = render(
      <AnchoredOverlayTestComponent onOpenCallback={mockOpenCallback} onCloseCallback={mockCloseCallback} />,
    )
    const anchor = anchoredOverlay.baseElement.querySelector('[aria-haspopup="true"]')!
    fireEvent.click(anchor)

    expect(mockOpenCallback).toHaveBeenCalledTimes(1)
    expect(mockOpenCallback).toHaveBeenCalledWith('anchor-click')
    expect(mockCloseCallback).toHaveBeenCalledTimes(0)
  })

  it('should call onOpen when the anchor activated by a key press', () => {
    const mockOpenCallback = vi.fn()
    const mockCloseCallback = vi.fn()
    const anchoredOverlay = render(
      <AnchoredOverlayTestComponent onOpenCallback={mockOpenCallback} onCloseCallback={mockCloseCallback} />,
    )
    const anchor = anchoredOverlay.baseElement.querySelector('[aria-haspopup="true"]')!
    fireEvent.keyDown(anchor, {key: ' '})

    expect(mockOpenCallback).toHaveBeenCalledTimes(1)
    expect(mockOpenCallback).toHaveBeenCalledWith('anchor-key-press')
    expect(mockCloseCallback).toHaveBeenCalledTimes(0)
  })

  it('should call onClose when the user clicks off of the overlay', () => {
    const mockOpenCallback = vi.fn()
    const mockCloseCallback = vi.fn()
    const anchoredOverlay = render(
      <AnchoredOverlayTestComponent
        initiallyOpen={true}
        onOpenCallback={mockOpenCallback}
        onCloseCallback={mockCloseCallback}
      />,
    )
    fireEvent.mouseDown(anchoredOverlay.baseElement)

    expect(mockOpenCallback).toHaveBeenCalledTimes(0)
    expect(mockCloseCallback).toHaveBeenCalledTimes(1)
    expect(mockCloseCallback).toHaveBeenCalledWith('click-outside')
  })

  it('should call onClose when the escape key is pressed', async () => {
    const mockOpenCallback = vi.fn()
    const mockCloseCallback = vi.fn()
    const anchoredOverlay = render(
      <AnchoredOverlayTestComponent
        initiallyOpen={true}
        onOpenCallback={mockOpenCallback}
        onCloseCallback={mockCloseCallback}
      />,
    )
    const overlay = await anchoredOverlay.findByRole('none')
    fireEvent.keyDown(overlay, {key: 'Escape'})

    expect(mockOpenCallback).toHaveBeenCalledTimes(0)
    expect(mockCloseCallback).toHaveBeenCalledTimes(1)
    expect(mockCloseCallback).toHaveBeenCalledWith('escape')
  })

  it('should render consistently when open', () => {
    const {container} = render(<AnchoredOverlayTestComponent initiallyOpen={true} />)
    expect(container).toMatchSnapshot()
  })

  it('should call onPositionChange when provided', () => {
    const mockPositionChangeCallback = vi.fn(({position}: {position: AnchorPosition}) => position)
    const anchoredOverlay = render(
      <AnchoredOverlayTestComponent initiallyOpen={true} onPositionChange={mockPositionChangeCallback} />,
    )
    const overlay = anchoredOverlay.baseElement.querySelector('[role="none"]')!
    fireEvent.keyDown(overlay, {key: 'Escape'})

    expect(mockPositionChangeCallback).toHaveBeenCalledTimes(1)
    expect(mockPositionChangeCallback).toHaveBeenCalledWith({
      position: {
        anchorAlign: 'start',
        anchorSide: 'outside-bottom',
        left: 0,
        top: 26.84375,
      },
    })
  })
})
