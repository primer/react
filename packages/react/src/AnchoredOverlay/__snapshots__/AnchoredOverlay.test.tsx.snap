// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AnchoredOverlay > should render consistently when open 1`] = `
<div>
  <div
    class="_BaseStyles_1n09d_56"
    data-color-mode="light"
    data-dark-theme="dark"
    data-light-theme="light"
    data-portal-root="true"
  >
    <button
      aria-describedby=":rd:-loading-announcement"
      aria-expanded="true"
      aria-haspopup="true"
      class="_ButtonBase_149hc_2"
      data-loading="false"
      data-no-visuals="true"
      data-size="medium"
      data-variant="default"
      id=":rd:"
      tabindex="0"
      type="button"
    >
      <span
        class="_ButtonContent_149hc_101"
        data-align="center"
        data-component="buttonContent"
      >
        <span
          class="_Label_149hc_128"
          data-component="text"
        >
          Anchor Button
        </span>
      </span>
    </button>
    <div
      id="__primerPortalRoot__"
      style="position: absolute; top: 0px; left: 0px;"
    >
      <div
        style="position: relative; z-index: 1;"
      >
        <div
          class="_Overlay_1a2ug_11"
          data-focus-trap="active"
          data-height-auto=""
          data-variant="anchored"
          data-visibility-visible=""
          data-width-auto=""
          role="none"
          style="left: 0px; top: 26.8438px;"
        >
          <span
            aria-hidden="true"
            class="sentinel"
            tabindex="0"
          />
          <button
            class="focus-visible"
            data-focus-visible-added=""
            tabindex="0"
            type="button"
          >
            Focusable Child
          </button>
          <span
            aria-hidden="true"
            class="sentinel"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
